packages/application-generic/src/usecases/trigger-multicast/trigger-multicast.usecase.ts

```diff
@@ -1,139 +1,207 @@
+      const getTopicDistinctSubscribersGenerator =
+        await this.topicSubscribersRepository.getTopicDistinctSubscribers({
+          query: {
+            _organizationId: organizationId,
+            _environmentId: environmentId,
+            topicIds: topicIds,
+            excludeSubscribers: singleSubscriberIds,
+          },
+          batchSize: subscriberFetchBatchSize,
+        });
 
-  @Instrument()
-  private async validateSubscriberIdProperty(
-    to: ISubscribersSource[]
-  ): Promise<boolean> {
-    for (const subscriber of to) {
-      const subscriberIdExists =
-        typeof subscriber === 'string' ? subscriber : subscriber.subscriberId;
-
-      if (Array.isArray(subscriberIdExists)) {
-        throw new ApiException(
-          'subscriberId under property to is type array, which is not allowed please make sure all subscribers ids are strings'
-        );
+      for await (const topicSubscriber of getTopicDistinctSubscribersGenerator) {
+        const externalSubscriberId = topicSubscriber._id;
+
+        if (actor && actor.subscriberId === externalSubscriberId) {
+          continue;
+        }
+
+        subscribersList.push({ subscriberId: externalSubscriberId });
+
+        if (subscribersList.length === subscriberFetchBatchSize) {
+          await this.sendToProcessSubscriberService(
+            command,
+            subscribersList,
+            SubscriberSourceEnum.TOPIC
+          );
+
+          subscribersList = [];
+        }
       }
 
-      if (!subscriberIdExists) {
-        throw new ApiException(
-          'subscriberId under property to is not configured, please make sure all subscribers contains subscriberId property'
+      if (subscribersList.length > 0) {
+        await this.sendToProcessSubscriberService(
+          command,
+          subscribersList,
+          SubscriberSourceEnum.TOPIC
         );
       }
     }
-
-    return true;
   }
 
-  @Instrument()
-  private async getProviderId(
+  private async getTopicsByTopicKeys(
+    organizationId: string,
     environmentId: string,
-    channelType: ChannelTypeEnum
-  ): Promise<ProvidersIdEnum> {
-    const integration = await this.integrationRepository.findOne(
+    topicKeys: Set<string>
+  ): Promise<TopicEntity[]> {
+    return await this.topicRepository.find(
       {
+        _organizationId: organizationId,
         _environmentId: environmentId,
-        active: true,
-        channel: channelType,
+        key: { $in: Array.from(topicKeys) },
       },
-      'providerId'
+      '_id'
     );
+  }
 
-    return integration?.providerId as ProvidersIdEnum;
+  private validateTopicExist(topics: TopicEntity[], topicKeys: Set<string>) {
+    if (topics.length !== topicKeys.size) {
+      topicKeys.forEach((topicKey) => {
+        if (!topics.find((topic) => topic.key === topicKey)) {
+          throw new NotFoundException(
+            `Topic with key ${topicKey} not found in current environment`
+          );
+        }
+      });
```

Comment:
True thanks for bringing it up.

I created it like this because I tried to mimic the old flow we had, but I agree that now we would benefit more from informative error.